surge.go

```diff
@@ -1,736 +1,279 @@
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [1]byte{0}
-		if valOf.Bool() {
-			bs[0] = 1
-		}
-		n, err := w.Write(bs[:])
-		return m - n, err
+		return MarshalBool(v.Bool(), buf, len(buf))
 
-	case reflect.Int8:
-		if m < 1 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [1]byte{byte(valOf.Int())}
-		n, err := w.Write(bs[:])
-		return m - n, err
+	case reflect.Uint8:
+		return MarshalU8(uint8(v.Uint()), buf, rem)
+	case reflect.Uint16:
+		return MarshalU16(uint16(v.Uint()), buf, rem)
+	case reflect.Uint32:
+		return MarshalU32(uint32(v.Uint()), buf, rem)
+	case reflect.Uint, reflect.Uint64:
+		return MarshalU64(uint64(v.Uint()), buf, rem)
 
+	case reflect.Int8:
+		return MarshalI8(int8(v.Int()), buf, rem)
 	case reflect.Int16:
-		if m < 2 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [2]byte{}
-		binary.BigEndian.PutUint16(bs[:], uint16(valOf.Int()))
-		n, err := w.Write(bs[:])
-		return m - n, err
-
+		return MarshalI16(int16(v.Int()), buf, rem)
 	case reflect.Int32:
-		if m < 4 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [4]byte{}
-		binary.BigEndian.PutUint32(bs[:], uint32(valOf.Int()))
-		n, err := w.Write(bs[:])
-		return m - n, err
-
+		return MarshalI32(int32(v.Int()), buf, rem)
 	case reflect.Int64:
-		if m < 8 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [8]byte{}
-		binary.BigEndian.PutUint64(bs[:], uint64(valOf.Int()))
-		n, err := w.Write(bs[:])
-		return m - n, err
-
-	case reflect.Uint8:
-		if m < 1 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [1]byte{byte(valOf.Uint())}
-		n, err := w.Write(bs[:])
-		return m - n, err
+		return MarshalI64(int64(v.Int()), buf, rem)
 
-	case reflect.Uint16:
-		if m < 2 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [2]byte{}
-		binary.BigEndian.PutUint16(bs[:], uint16(valOf.Uint()))
-		n, err := w.Write(bs[:])
-		return m - n, err
-
-	case reflect.Uint32:
-		if m < 4 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [4]byte{}
-		binary.BigEndian.PutUint32(bs[:], uint32(valOf.Uint()))
-		n, err := w.Write(bs[:])
-		return m - n, err
-
-	case reflect.Uint64:
-		if m < 8 {
-			return m, ErrMaxBytesExceeded
-		}
-		bs := [8]byte{}
-		binary.BigEndian.PutUint64(bs[:], uint64(valOf.Uint()))
-		n, err := w.Write(bs[:])
-		return m - n, err
+	case reflect.Float32:
+		return MarshalF32(float32(v.Float()), buf, rem)
+	case reflect.Float64:
+		return MarshalF64(float64(v.Float()), buf, rem)
 
 	case reflect.Array:
-		err := error(nil)
-		len := valOf.Len()
-		for i := 0; i < len; i++ {
-			m, err = Marshal(w, valOf.Index(i).Interface(), m)
-			if err != nil {
-				return m, err
-			}
-		}
-		return m, nil
-
+		return marshalReflectedArray(v, buf, rem)
 	case reflect.Slice:
-		len := valOf.Len()
-		m, err := Marshal(w, uint32(len), m)
-		if err != nil {
-			return m, err
-		}
-		for i := 0; i < len; i++ {
-			m, err = Marshal(w, valOf.Index(i).Interface(), m)
-			if err != nil {
-				return m, err
-			}
-		}
-		return m, nil
-
+		return marshalReflectedSlice(v, buf, rem)
```

Comment:
Why is the reflected version used here? Can't we use `v.String()` and use the typed function like for the primitive types in the cases above?